temporalio/worker/_workflow_instance.py

```diff
@@ -429,15 +429,20 @@ def activate(
                     f"Failed converting activation exception: {inner_err}"
                 )
 
-        def is_completion(command):
+        def is_non_cancellation_completion(command):
             return (
                 command.HasField("complete_workflow_execution")
                 or command.HasField("continue_as_new_workflow_execution")
                 or command.HasField("fail_workflow_execution")
-                or command.HasField("cancel_workflow_execution")
             )
 
-        if any(map(is_completion, self._current_completion.successful.commands)):
+        # We do also warn in the case of workflow cancellation, but this is done
+        # when handling the workflow cancellation, since we also cancel update
+        # handlers at that time.
```

Comment:
IMO it should be at workflow completion time where we warn always, regardless of reason for completion. This includes cancellation. Users should ensure that their handlers complete even during cancellation, but that doesn't need to be checked until the actual workflow completion is being recorded (primary workflow method could be swallowing cancellation, throwing some other error, etc).